### PR TITLE
Add Module Workers article

### LINKED ISSUE #1915: content: ES Modules for Dedicated Workers
**Author**: @developit 
**Reviewer**: @jpmedley 

**Target publish date:** 2019-12-19
- [x] Check this box if this is a hard deadline.

- [ ] Check this box if you'd like your reviewer to fix grammar and punctuation issues for you.

**Process**

Check out the [web.dev writing process](https://web.dev/handbook/quick-start/).

When you're ready, make a copy of [this template](https://docs.google.com/document/d/1lgaNIEnXZf-RB8_p9RK22QEgpXJqnu77pLWVWVy4nuw/edit)
and fill in the proposal and outline sections. When you're finished, drop a link
to the doc in the 1st step below and check the box. A member of the web.dev team
will assign this ticket to themselves and work as your reviewer to help you
complete the remaining steps.

- [ ] 1. I've done my proposal <add a link to your proposal doc>
- [ ] 2. Proposal approved — _reviewer checks this_
- [x] 3. I've added a draft to my proposal doc https://docs.google.com/document/d/1dAQbe40l_-38jb_TMddpu76lJ7neOZvYOMxAH6LB4hI/edit#heading=h.gwohrbbkk7vp
- [ ] 4. Draft approved — _reviewer checks this_
- [ ] 5. I've submitted a pull request #2006
- [ ] 6. Pull request approved — _reviewer checks this_

### PR DESCRIPTION
**Pull request by developit**
Fixes #1915.

Changes proposed in this pull request:

- Adds Module Workers article.

### CODE REVIEW

File: src/site/content/en/blog/module-workers/index.md [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+##### greet.js:
+
+```js
+import greetings from './data.js';
+export function sayHello() {
+  return greetings.hello;
+}
+```
+
+To ensure great performance, the old `importScripts()` API is not available within Module Workers. Switching Workers to use JS Modules means all code is loaded in [strict mode](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode), and HTML-style comments are not supported. Another notable change is that the value of `this` in the top-level scope of a JS Module is `undefined`, whereas in Classic Workers the value was the worker's global scope. Fortunately, there has always been a `self` global that provides a reference to the global scope - it's available in all types of Workers including Service Workers, as well as in the DOM.
```
Comment (jeffposnick) [inline src/site/content/en/blog/module-workers/index.md]: What's the significance of "HTML-style comments" not being supported? I'm having a hard time imagining how HTML-style comments would have been used in an classic worker—was that really a thing?

Comment (developit) [reply src/site/content/en/blog/module-workers/index.md]: I believe they were supported, though I can't imagine anyone has ever used them. Maybe for IE conditional comments?

Comment (jeffposnick) [reply src/site/content/en/blog/module-workers/index.md]: Weird.

I'd consider just dropping mention of that, or move it to an aside or something? For an edge case, it's needlessly puzzling.

File: src/site/content/en/blog/module-workers/index.md [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+  background threads. Module Workers bring modern JavaScript
+tags:
+  - post
+  - web-workers
+  - js-modules
+---
+
+<!--lint disable no-heading-punctuation heading-increment -->
+
+JavaScript is single-threaded, which means only one operation can be performed at a time. This is
```
Comment (jpmedley) [inline src/site/content/en/blog/module-workers/index.md]: Please use active voice. Change 'only one operation can be performed at a time' to 'it can only perform one operation at a time.'

File: src/site/content/en/blog/module-workers/index.md [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+intuitive and works well for lots of cases on the web, but can become problematic when we need to
+do heavy lifting tasks like data processing, parsing, computation or analysis. As more and more
+complex applications are delivered on the Web Platform, there's an increased need for
+multi-threaded processing.
+
+On the Web Platform, the main primitive for threading and parallelism is the [Web
```
Comment (jpmedley) [inline src/site/content/en/blog/module-workers/index.md]: 'Web Platform' should be lower case.

File: src/site/content/en/blog/module-workers/index.md [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+```js
+addEventListener('message', e => {
+  if (e.data === 'hello') {
+    postMessage('world');
+  }
+});
+```
+
+
+Web Workers have actually been available in most browsers for over 10 years. While that means they
```
Comment (jpmedley) [inline src/site/content/en/blog/module-workers/index.md]: '10' should be 'ten'.

File: src/site/content/en/blog/module-workers/index.md [hunk 5]
```diff
@@ -0,0 +1,229 @@
+---
+title: "Threading the Web with Module Workers"
```
Comment (jpmedley) [inline src/site/content/en/blog/module-workers/index.md]: Web.dev uses sentence case for titles (unlike what your writing teachers taught you). This should be 'Threading the web with module workers'.

Please check your section titles throughout.

File: src/site/content/en/blog/module-workers/index.md [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+```
+
+
+Web Workers have actually been available in most browsers for over 10 years. While that means they
+have excellent browser support and are well-optimized, it also means they long predate JavaScript
+Modules. Since there was no module system when Workers were designed, the API for loading code into
+a Worker and composing scripts has remained similar to the synchronous script loading approaches
+common in 2009.
+
+## History: Classic Workers
```
Comment (jpmedley) [inline src/site/content/en/blog/module-workers/index.md]: 'History: classic workers'

File: src/site/content/en/blog/module-workers/index.md [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+Script](https://html.spec.whatwg.org/multipage/webappapis.html#classic-script) URL, which is
+relative to the document URL. It immediately returns a reference to the new Worker instance, which
+exposes a messaging interface as well as a `terminate()` method that immediately stops and destroys
+the worker.
+
+```js
+const worker = new Worker('worker.js');
+```
+
+An `importScripts()` function is available within Web Workers for loading additional code, but it
```
Comment (jpmedley) [inline src/site/content/en/blog/module-workers/index.md]: 'web workers' should actually be lower case. You would use upper case for referring to the API by name or if referring to the interface explicitly. Please check this throughout.

File: src/site/content/en/blog/module-workers/index.md [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+##### greet.js:
+
+```js
+// global to the whole worker
+function sayHello() {
+  return 'world';
+}
+```
+
+For this reason, using Web Workers has historically imposed an outsized effect on the architecture
```
Comment (jpmedley) [inline src/site/content/en/blog/module-workers/index.md]: 'using Web Workers has' 

You can shorten this to 'web workers have'.

File: src/site/content/en/blog/module-workers/index.md [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+Worker constructor accepts a new `{type:"module"}` option, which changes script loading and
+execution to match `<script type="module">`.
+
+```js
+const worker = new Worker('worker.js', {
+  type: 'module'
+});
+```
+
+Since Module Workers are JS Modules, they can use standard JavaScript import and export statements.
```
Comment (jpmedley) [inline src/site/content/en/blog/module-workers/index.md]: Always spell out JavaScript. Please check throughout.

File: src/site/content/en/blog/module-workers/index.md [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+As with all JS Modules, dependencies are only executed once in a given context (main thread,
+worker, etc), and all future imports reference the already-executed module instance. The loading
```
Comment (jpmedley) [inline src/site/content/en/blog/module-workers/index.md]: 'etc.' needs a period.

File: src/site/content/en/blog/module-workers/index.md [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+and execution of JS Modules is also optimized by browsers - because a Module's dependencies can be
```
Comment (jpmedley) [inline src/site/content/en/blog/module-workers/index.md]: * Remove the dash.
* Make "module's" lower case.

File: src/site/content/en/blog/module-workers/index.md [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+loaded prior to the Module being executed, entire module trees can be loaded in parallel. Module
```
Comment (jpmedley) [inline src/site/content/en/blog/module-workers/index.md]: 'Module' should be lower case.

File: src/site/content/en/blog/module-workers/index.md [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+loading also caches parsed code, which means modules that are used on the main thread and in a
+Worker only need to be parsed once.
+
+Moving to JS Modules also enables the use of [dynamic
+import](https://v8.dev/features/dynamic-import) for lazy-loading code without blocking execution of
+the worker. Dynamic Import is much more explicit than using importScripts to load dependencies,
```
Comment (jpmedley) [inline src/site/content/en/blog/module-workers/index.md]: * 'Import' to 'import' (lower case)
* importScripts needs to be `importScripts()` (formatting and parens added)

File: src/site/content/en/blog/module-workers/index.md [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+##### greet.js:
+
+```js
+import greetings from './data.js';
+export function sayHello() {
+  return greetings.hello;
+}
+```
+
+To ensure great performance, the old `importScripts()` API is not available within Module Workers.
```
Comment (jpmedley) [inline src/site/content/en/blog/module-workers/index.md]: 'Module Workers' --> 'module workers'

File: src/site/content/en/blog/module-workers/index.md [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+Switching Workers to use JS Modules means all code is loaded in [strict
```
Comment (jpmedley) [inline src/site/content/en/blog/module-workers/index.md]: * 'Workers' -- > 'workers'
* 'JS Modules' --> 'JavaScript modules'

File: src/site/content/en/blog/module-workers/index.md [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+mode](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode), and
+HTML-style comments are not supported. Another notable change is that the value of `this` in the
+top-level scope of a JS Module is `undefined`, whereas in Classic Workers the value was the
```
Comment (jpmedley) [inline src/site/content/en/blog/module-workers/index.md]: * 'JS Module' --> 'JavaScript Module'
* 'Classic Workers' --> 'classic workers'

File: src/site/content/en/blog/module-workers/index.md [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+worker's global scope. Fortunately, there has always been a `self` global that provides a reference
+to the global scope - it's available in all types of Workers including Service Workers, as well as
```
Comment (jpmedley) [inline src/site/content/en/blog/module-workers/index.md]: "It's" is the start of a new sentence.

File: src/site/content/en/blog/module-workers/index.md [hunk 18]
```diff
@@ -0,0 +1,229 @@
...
+
+Preloaded modules can also be used by both the main thread and Module Workers. This is useful for
+modules that are imported in both contexts, or in cases where it's not possible to know in advance
+whether a module will be used on the main thread or in a Worker.
+
+Previously, the options available for preloading Web Worker scripts were limited and not
+necessarily reliable. Classic Workers had their own "worker" resource type for preloading, but no
+browsers implemented `<link rel="preload" as="worker">`. As a result, the primary technique
+available for preloading Worker scripts was to use `<link rel="prefetch">`, which relied entirely
+on the HTTP cache. When used in combination with caching headers were set up correctly, this made
```
Comment (jpmedley) [inline src/site/content/en/blog/module-workers/index.md]: The beginning of this sentence runs off the rails. I don't have a suggestion because I'm not sure what you're trying to say.

File: src/site/content/en/blog/module-workers/index.md [hunk 19]
```diff
@@ -0,0 +1,229 @@
...
+it possible to avoid `new Worker()` having to wait to download the Worker script, but did not
```
Comment (jpmedley) [inline src/site/content/en/blog/module-workers/index.md]: Change '`new Worker()`' to 'worker instantiation'. It's a little more formal, which is more appropriate for this content channel.

File: src/site/content/en/blog/module-workers/index.md [hunk 20]
```diff
@@ -0,0 +1,229 @@
...
+  type: 'module'
+});
+```
+
+Now that the specification has been updated, browsers are beginning to implement the new behavior.
+This takes time because there are some extra complications associated with bringing JS Modules to
+Service Worker. Service Worker registration needs to [compare imported scripts with their previous
+cached versions](https://chromestatus.com/feature/6533131347689472) when determining whether to
+trigger an update, and this needs to be implemented for JS Modules when used for Service Workers.
+Also, Service Worker needs to be able to [bypass the
```
Comment (jpmedley) [inline src/site/content/en/blog/module-workers/index.md]: 'Service Worker needs' to 'Service workers need'

File: src/site/content/en/blog/module-workers/index.md [hunk 21]
```diff
@@ -0,0 +1,229 @@
+---
+title: "Threading the Web with Module Workers"
+subhead: |
+  Moving heavy lifting into background threads is now easier with JS Modules in Workers.
```
Comment (jpmedley) [inline src/site/content/en/blog/module-workers/index.md]: Is 'JS Modules' the official name of the spec or is it actually 'JavaScript Modules'. Use the exact name and spelling

Also if you're not using the spec name it should be . 'JavaScript modules in workers.' (Notice the capitalization.)

File: src/site/content/en/blog/module-workers/index.md (hunk 7 above)
Comment (jpmedley) [reply src/site/content/en/blog/module-workers/index.md]: There are other capitalization things I noted. I corrected them in some places but not others.

File: src/site/content/en/blog/module-workers/index.md (hunk 21 above)
Comment (developit) [reply src/site/content/en/blog/module-workers/index.md]: "JavaScript Modules" is the official name

File: src/site/content/en/blog/module-workers/index.md (hunk 5 above)
Comment (developit) [reply src/site/content/en/blog/module-workers/index.md]: Sounds good - only concern is that "Workers" is typically capitalized, because it's a reference to the feature named Web Workers. I was taking "Module Workers" to be a feature name here, though that is admittedly something I've only seen done colloquially - there is no specification called Module Workers, only a JS Modules addendum to the Web Workers specification.

File: src/site/content/en/blog/module-workers/index.md [hunk 22]
```diff
@@ -0,0 +1,229 @@
...
+date: 2019-12-17
+hero: hero.jpg
+alt: Module Workers
+authors:
+  - developit
+description: |
+  JavaScript is single-threaded, but Web Workers can be used to move expensive operations into
+  background threads. Module Workers bring modern JavaScript
```
[the remark this answers is not in the document]
Comment (developit) [reply src/site/content/en/blog/module-workers/index.md]: missing indeed! apologies.

File: src/site/content/en/blog/module-workers/index.md (hunk 18 above)
Comment (developit) [reply src/site/content/en/blog/module-workers/index.md]: woah yeah, let me finish that thought.

**Comment by developit** [comment]
Thank you so much for the grammar fixes and catching my mistakes @jpmedley! I believe I've corrected everything now.

File: src/site/content/en/blog/module-workers/index.md [hunk 23]
```diff
@@ -0,0 +1,232 @@
...
+To ensure great performance, the old `importScripts` API is not available within module workers.
+Switching workers to use JavaScript Modules means all code is loaded in [strict
+mode](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode). Another
+notable change is that the value of `this` in the top-level scope of a JavaScript Module is
+`undefined`, whereas in classic workers the value is the worker's global scope. Fortunately, there
+has always been a `self` global that provides a reference to the global scope. It's available in
+all types of workers including service workers, as well as in the DOM.
+
+{% Aside %}
+Module workers also remove support for HTML-style comments.
```
Comment (developit) [inline src/site/content/en/blog/module-workers/index.md]: I worry the UI treatment for Aside is a little too much for this. It's more of a lark, since very few people will even realize HTML comments were supported in classic workers. I had thought of something like this but worried it was too conversational:

> Module workers also remove support for HTML-style comments. Did you know you could use HTML comments in web worker scripts?

File: src/site/content/en/blog/module-workers/index.md [hunk 24]
```diff
@@ -0,0 +1,232 @@
...
+  });
+</script>
+```
+
+Preloaded modules can also be used by both the main thread and module workers. This is useful for
+modules that are imported in both contexts, or in cases where it's not possible to know in advance
+whether a module will be used on the main thread or in a worker.
+
+Previously, the options available for preloading web worker scripts were limited and not
+necessarily reliable. Classic workers had their own "worker" resource type for preloading, but no
```
Comment (developit) [inline src/site/content/en/blog/module-workers/index.md]: @jpmedley this was the past-tense paragraph I was lamenting. We're not removing classic workers, we're just adding a new modern alternative to them.

File: src/site/content/en/blog/module-workers/index.md (hunk 23 above)
Comment (jeffposnick) [reply src/site/content/en/blog/module-workers/index.md]: Or just drop it? The only reason I bring this up is because as soon as I read this I fell into a wormhole trying to figure out how HTML comments related to classic workers, and that could take someone out of the flow of digesting the actual information you want them to focus on.

File: src/site/content/en/blog/module-workers/index.md (hunk 24 above)
Comment (jpmedley) [reply src/site/content/en/blog/module-workers/index.md]: 'Classic workers had their own' --> 'Classic workers once had their own'

File: src/site/content/en/blog/module-workers/index.md (hunk 23 above)
Comment (kaycebasques) [reply src/site/content/en/blog/module-workers/index.md]: > Module workers also remove support for HTML-style comments. Did you know you could use HTML comments in web worker scripts?

A little fun and humor is welcome. I also didn't know that this was supported and was amused to find out. This little tidbit didn't personally distract me from the main flow too much. But @jeffposnick's rationale is totally on point: we avoid side information in order to help readers stay focused on the most important main message.

File: src/site/content/en/blog/module-workers/index.md [hunk 25]
```diff
@@ -0,0 +1,232 @@
...
+
+<div class="glitch-embed-wrap" style="height: 480px; width: 100%;">
+  <iframe
+    src="https://glitch.com/embed/#!/embed/worker-preloading?previewSize=100&attributionHidden=true"
+    alt="worker-preloading on Glitch"
+    style="height: 100%; width: 100%; border: 0;">
+  </iframe>
+</div>
+
+## What about Service Worker?
```
Comment (kaycebasques) [inline src/site/content/en/blog/module-workers/index.md]:
```suggestion
## What about service worker?
```

Nit: I can't find our own personal guidance on this, but service worker should be lower case. The spec backs this convention up: https://www.w3.org/TR/service-workers/#dfn-service-worker

File: src/site/content/en/blog/module-workers/index.md [hunk 26]
```diff
@@ -0,0 +1,232 @@
...
+
+The Service Worker specification [has already been
```
Comment (kaycebasques) [inline src/site/content/en/blog/module-workers/index.md]:
```suggestion
The service worker specification [has already been
```
Nit

File: src/site/content/en/blog/module-workers/index.md (hunk 24 above)
Comment (kaycebasques) [reply src/site/content/en/blog/module-workers/index.md]: @developit I think you have already made it clear that you're not removing classic workers because this is how you introduce the feature:

> A new mode for web workers with the ergonomics and performance benefits of [JavaScript
modules](https://v8.dev/features/modules) is shipping in Chrome 80, called module workers. 

`mode` gets the idea across clearly.

File: src/site/content/en/blog/module-workers/index.md [hunk 27]
```diff
@@ -0,0 +1,232 @@
...
+description: |
+  Module workers make it easy to unblock the main thread by moving expensive code to a background
+  thread while keeping the ergonomic and performance benefits of standard JavaScript modules.
+tags:
+  - post
+  - web-workers
+  - js-modules
+---
+
+<!--lint disable no-heading-punctuation heading-increment -->
```
Comment (kaycebasques) [inline src/site/content/en/blog/module-workers/index.md]:
```suggestion
```

Presumably this is instructing the linter to ignore the checks around headings levels. That's actually something that we must fix before publishing this.

**Comment by developit** [comment]
Sounds good. I didn't know I had write access!
